code/network/lib/Loot/Network/ZMQ/Client.hs

```diff
@@ -123,15 +160,43 @@ instance T.Show InternalRequest where
     show (IRRegister cId _ _ _) = "IRRegister " <> show cId
     show (IRReconnect nids)     = "IRReconnect" <> show nids
 
-applyUpdatePeers ::
-       Set ZTNodeId -> ZTUpdatePeersReq -> (Set ZTNodeId, Set ZTNodeId)
-applyUpdatePeers peers (UpdatePeersReq {..}) = do
-    let both = _uprDel `Set.intersection` _uprAdd
-    let add' = (_uprAdd `Set.difference` both) `Set.difference` peers
-    let del' = (_uprDel `Set.difference` both) `Set.intersection` peers
-    (add',del')
+-- Get hashmap keys as a hashset.
+hmKeys :: HashMap k a -> HashSet k
+hmKeys = HSet.fromMap . fmap (const ())
 
-newtype CliRequestQueue = CliRequestQueue { unCliRequestQueue :: TQueue InternalRequest }
+-- Reformulates peers update request in terms of current peer info
+-- var, returning the set of add/del nodes. Doesn't change PIV itself.
+applyUpdatePeers ::
+       PeersInfoVar
+    -> ZTUpdatePeersReq
+    -> STM ([ZTNodeId], [ZTNodeId])
+applyUpdatePeers piv UpdatePeersReq{..} = do
+    -- Normalising the request
+    let both = L.nub $ _uprDel `L.intersect` _uprAdd
+    let add' = L.nub _uprAdd L.\\ both
+    let del' = L.nub _uprDel L.\\ both
```

Comment:
I don't see how that would follow. `where` case is not created to merely decouple the function logic, it also does more (it can loop, it can be called several times). Comments are just comments.